ledger/participant-integration-api/src/main/scala/platform/store/dao/events/ContractLoader.scala

```diff
@@ -0,0 +1,155 @@
+// Copyright (c) 2023 Digital Asset (Switzerland) GmbH and/or its affiliates. All rights reserved.
+// SPDX-License-Identifier: Apache-2.0
+
+package com.daml.platform.store.dao.events
+
+import akka.stream.{BoundedSourceQueue, Materializer, QueueOfferResult}
+import akka.stream.scaladsl.{Keep, Sink, Source}
+import com.daml.error.{ContextualizedErrorLogger, DamlContextualizedErrorLogger}
+import com.daml.error.definitions.LedgerApiErrors
+import com.daml.ledger.offset.Offset
+import com.daml.ledger.resources.ResourceOwner
+import com.daml.lf.value.Value.ContractId
+import com.daml.logging.LoggingContext
+import com.daml.metrics.api.MetricsContext
+import com.daml.metrics.{InstrumentedGraph, Metrics}
+import com.daml.platform.indexer.parallel.BatchN
+import com.daml.platform.store.backend.ContractStorageBackend
+import com.daml.platform.store.dao.DbDispatcher
+
+import scala.concurrent.{ExecutionContext, Future, Promise}
+import scala.util.{Failure, Success}
+
+trait Loader[KEY, VALUE] {
+
+  def load(key: KEY)(implicit loggingContext: LoggingContext): Future[Option[VALUE]]
+
+}
+
+class AkkaStreamParallelBatchedLoader[KEY, VALUE](
+    batchLoad: Seq[(KEY, LoggingContext)] => Future[Map[KEY, VALUE]],
+    createQueue: () => Source[(KEY, LoggingContext, Promise[Option[VALUE]]), BoundedSourceQueue[
+      (KEY, LoggingContext, Promise[Option[VALUE]])
+    ]],
+    maxBatchSize: Int,
+    parallelism: Int,
+)(implicit executionContext: ExecutionContext, materializer: Materializer)
+    extends Loader[KEY, VALUE] {
+  private implicit val contextualizedErrorLogger: ContextualizedErrorLogger =
+    DamlContextualizedErrorLogger.forClass(this.getClass)
+
+  private val (queue, done) = createQueue() // TODO maybe set up autorestart?
```

Comment:
I am not sure this is needed. normally there is no reason for this stream to terminate...WDYT @tudor-da

Reply:
what if `batchLoad` throws?

Reply:
like before the async part? or what do you mean?
if the future fails we map that to success a bit down lower

Reply:
But `batchLoad` is not inside a Future, only its result is. If it throws for some unhandled case, it can kill the Source

Reply:
will do defer there :+1: